pkg/ipam-controller/controllers/configmap/configmap.go

```diff
@@ -0,0 +1,127 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/allocator"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/config"
+	"github.com/Mellanox/nvidia-k8s-ipam/pkg/ipam-controller/selector"
+)
+
+// ConfigMapReconciler reconciles ConfigMap objects
+type ConfigMapReconciler struct {
+	Allocator          *allocator.Allocator
+	Selector           *selector.Selector
+	ConfigMapName      string
+	ConfigMapNamespace string
+
+	ConfigEventCh chan event.GenericEvent
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// Reconcile contains logic to sync Node objects
+func (r *ConfigMapReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLog := log.FromContext(ctx)
+	if req.Name != r.ConfigMapName || req.Namespace != r.ConfigMapNamespace {
+		// this should never happen because of the watcher configuration
+		return ctrl.Result{}, nil
+	}
+
+	cfg := &corev1.ConfigMap{}
+	err := r.Client.Get(ctx, req.NamespacedName, cfg)
+	if err != nil {
+		reqLog.Error(err, "failed to get ConfigMap object from the cache")
+		return ctrl.Result{}, err
+	}
+
+	confData, exist := cfg.Data[config.ConfigMapKey]
+	if !exist {
+		reqLog.Error(nil, fmt.Sprintf("invalid configuration: ConfigMap %s doesn't contain %s key",
+			r.ConfigMapNamespace, config.ConfigMapKey))
+		return ctrl.Result{}, nil
+	}
+
+	controllerConfig := &config.Config{}
+	if err := json.Unmarshal([]byte(confData), controllerConfig); err != nil {
+		reqLog.Error(err, fmt.Sprintf("invalid configuration: ConfigMap %s contains invalid JSON",
+			config.ConfigMapKey))
+		return ctrl.Result{}, nil
+	}
+
+	if err := controllerConfig.Validate(); err != nil {
+		reqLog.Error(err, fmt.Sprintf("invalid configuration: ConfigMap %s contains invalid config",
+			config.ConfigMapKey))
+		return ctrl.Result{}, nil
+	}
+
+	r.Selector.Update(controllerConfig.NodeSelector)
+
+	allocatorPools := make([]allocator.AllocationConfig, 0, len(controllerConfig.Pools))
+	for pName, p := range controllerConfig.Pools {
+		// already validated by Validate function
+		_, subnet, _ := net.ParseCIDR(p.Subnet)
+		gateway := net.ParseIP(p.Gateway)
+		allocatorPools = append(allocatorPools, allocator.AllocationConfig{
+			PoolName:         pName,
+			Subnet:           subnet,
+			Gateway:          gateway,
+			PerNodeBlockSize: p.PerNodeBlockSize,
+		})
+	}
+
+	if r.Allocator.IsConfigured() {
+		r.Allocator.Configure(ctx, allocatorPools)
+	} else {
+		nodeList := &corev1.NodeList{}
+		if err := r.Client.List(ctx, nodeList); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.Allocator.ConfigureAndLoadAllocations(ctx, allocatorPools, nodeList.Items)
+	}
+
+	// config updated, trigger sync for all nodes
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList); err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, n := range nodeList.Items {
```

Comment:
this will trigger the node controller Reconcile to sync again all nodes right?

Reply:
yes

Reply:
Code is designed to be able to handle each node object individually, and it is able to handle multiple node objects concurrently.
The Node controller currently runs with a single worker to avoid possible concurrency-related issues. We can increase the count of workers later when we will have good enough test coverage.